# Concurrency and Parallelism
## Concurrency
**Definition**
Concurrency is the ability of a program to handle multiple tasks simultaneously (alternately and very quickly), although they don't necessarily run at the same time.
**Analogy**
Imagine a chef cooking three dishes. He or she switches between tasks rapidly: waiting for the water to boil, then chopping vegetables, then stirring the soup. They aren't done simultaneously, but they all run.
**Technical**
* Typically uses a single-core CPU
* Assisted by threading, asynchronous I/O, or coroutines (such as async and await in Python)
* Suitable for I/O-bound tasks (e.g., reading files, web scraping, database queries)
## Parallelism
**Definition**
Parallelism is the ability to execute multiple tasks completely concurrently, at the same time, using multiple CPU cores or multiple processors.
**Analogy**
Imagine three chefs in three kitchens, each cooking a single dish. All tasks run in parallel and simultaneously.
**Technical**
* Requires a multi-core CPU
* Uses parallel processing (multiprocessing, joblib, GPU processing)
* Suitable for CPU-bound tasks (e.g., heavy computing, image processing, training ML models)

"All parallelism is concurrent, but not all concurrency is parallel."

# Simple

This program demonstrates how tasks can be executed separately but synchronized so that the main program waits for each task to complete before continuing with the final execution.

In [1]:
import threading
import time

def task():
    print("Task started...")
    time.sleep(5)
    print("Task completed!")
    time.sleep(5)
    print("Task completed!")

# Create a thread
t = threading.Thread(target=task)

# Start a thread
t.start()

# Wait for a thread to complete
t.join()

print("All tasks completed!")

Task started...
Task completed!
Task completed!
All tasks completed!


This program illustrates how a task can run concurrently with the main program, allowing other parts of the program to continue running without having to wait for the task to complete.

In [2]:
import threading
import time

def task():
    print("Task started...")
    time.sleep(5)
    print("Task completed!")

# Create a thread
t = threading.Thread(target=task)

# Start a thread
t.start()

print("All tasks completed!")

Task started...
All tasks completed!
Task completed!


Tugas selesai!


## Example 1: Creating Threads for Parallel Execution

This code illustrates two activities running simultaneously using two separate threads: one printing the numbers 1 through 5, and the other printing the letters ‘a’ through ‘e’. Both run simultaneously, allowing the numbers and letters to appear alternately depending on the execution speed of each thread. The main program waits for both threads to complete before ending execution.

In [5]:
import threading
import time

def print_numbers():
    for i in range(1, 6):
        print(f"Number: {i}")
        time.sleep(1)

def print_letters():
    for letter in 'abcde':
        print(f"Letter: {letter}")
        time.sleep(1)

thread1 = threading.Thread(target=print_numbers)
thread2 = threading.Thread(target=print_letters)

thread1.start()
thread2.start()

thread1.join()
thread2.join()


Number: 1
Letter: a
Number: 2Letter: b

Letter: c
Number: 3
Letter: dNumber: 4

Letter: eNumber: 5



Explanation:
* In this example, there are two functions, print_numbers and print_letters, each simulating a task.
* We create two threads, thread1 and thread2, to run these functions concurrently.
* thread1.start() and thread2.start() start the threads.
* We use thread1.join() and thread2.join() to wait for the threads to complete.
* These threads run concurrently, printing numbers and letters, and sleeping for 1 second after printing each character.

## Example 2: Thread Safety with Shared Resources
This code demonstrates how two threads can work together to increment a shared variable (counter) 100,000 times each, while avoiding conflicts or errors due to concurrent access. By using locks, the code ensures that only one thread can modify the counter at a time, maintaining the accuracy of the final result. After both threads have finished executing, the program prints the final value of the counter, which should be 200,000 if all operations have been executed successfully without conflicts.

In [6]:
import threading

counter = 0
lock = threading.Lock()

def increment():
    global counter
    for _ in range(100000):
        with lock:
            counter += 1

thread1 = threading.Thread(target=increment)
thread2 = threading.Thread(target=increment)

thread1.start()
thread2.start()

thread1.join()
thread2.join()

print("Counter:", counter)


Counter: 200000


Explanation:\
This example illustrates a common problem in concurrent programming: shared resources.
We have a global counter variable that is incremented by multiple threads simultaneously.
To ensure thread safety, we use threading.Lock to protect the critical section where the counter is updated.
Each thread, thread1 and thread2, increments the counter 100,000 times.
After both threads complete, we print the final value of the counter. Using a lock ensures that the counter is updated safely.

# Intermediate

## Example 1: Producer-Consumer Scenario with Threads
This code illustrates a producer-consumer scenario using two threads and a queue structure (queue.Queue) for communication between the threads. The producer thread produces five items and places them in the queue, while the consumer thread retrieves items one by one from the queue and prints them to indicate that they have been "consumed." Once the producer is finished, it sends a special signal (None) to the consumer to indicate that no more items will be produced, allowing the consumer thread to terminate cleanly. This approach maintains synchronization and workflow between the two threads without conflicts.

In [3]:
import threading
import queue

def producer(q):
    for i in range(5):
        q.put(i)

def consumer(q):
    while True:
        item = q.get()
        if item is None:
            break
        print("Consumed:", item)

q = queue.Queue()
producer_thread = threading.Thread(target=producer, args=(q,))
consumer_thread = threading.Thread(target=consumer, args=(q,))

producer_thread.start()
consumer_thread.start()

producer_thread.join()
q.put(None) # Signal to consumer to stop
consumer_thread.join()

Consumed: 0
Consumed: 1
Consumed: 2
Consumed: 3
Consumed: 4


Explanation:
* In this example, we have two threads representing a producer and a consumer.
* The producer thread, producer_thread, produces and adds items to a shared queue.
* The consumer thread, consumer_thread, consumes items from the queue.
* The consumer thread continues to consume items until it receives a "stop" (None) signal from the producer.
* The use of queues ensures concurrency safety and allows secure communication between producers and consumers.

## Example 2: Thread Pool for Task Parallelism
This code demonstrates the use of ThreadPoolExecutor from the concurrent.futures module to execute functions in parallel using threads. The square function calculates the square of a number after a one-second delay. The list of numbers [1, 2, 3, 4, 5] is processed simultaneously by up to three parallel threads. This approach results in a faster overall execution compared to running them individually, as multiple tasks are running concurrently. After all the results are calculated, the program prints a list of the squared results.

In [8]:
from concurrent.futures import ThreadPoolExecutor
import time

def square(x): 
    time.sleep(1) 
    return x*x

data = [1, 2, 3, 4, 5]
with ThreadPoolExecutor(max_workers=3) as executor: 
    result = list(executor.map(square, data))

print("Result:", result)

Result: [1, 4, 9, 16, 25]


Explanation:
In this example, we introduce ThreadPoolExecutor from the concurrent.futures module, which provides a convenient way to manage and execute tasks concurrently.
We define a square function that simulates a task by squaring a number and sleeping for 1 second.
We create a list of data and use the executor.map method for concurrent execution.

# Advanced

## Example 1: Concurrent Execution with ThreadPoolExecutor
This code is an implementation of concurrent programming using ThreadPoolExecutor from the concurrent.futures module to execute functions in parallel with threads. The square function will calculate the square of each number in a data list containing the numbers 0 through 99, with a one-second pause between each execution. Because max_workers=3, only three tasks will be executed concurrently by the three active threads. Although each calculation takes time, this approach speeds up the overall process compared to performing them sequentially. Once all squares have been calculated in parallel, the results are aggregated and displayed as a list.

In [9]:
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
import time

def square(x):
    time.sleep(1)
    return x * x

data = [i for i in range(100)]
with ThreadPoolExecutor(max_workers=3) as executor:
    result = list(executor.map(square, data))

print("Result (Thread):", result)


Result (Thread): [0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801]


This code demonstrates parallel processing using ThreadPoolExecutor with a large number of workers (max_workers=100). The square function calculates the square of a number in a data list containing the numbers 0 through 99, with a one-second pause between each processing step to simulate a time-consuming process. Because the number of threads is 100 equals the number of elements in the data, all tasks can be processed simultaneously, so the total execution time is close to the execution time of a single function (about 1 second), rather than 100 seconds as with sequential processing. This approach demonstrates the power of concurrency in maximizing thread utilization for increased efficiency, although in practice, too many threads can put stress on the system depending on load and resources.

In [10]:
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
import time

def square(x):
    time.sleep(1)
    return x * x

data = [i for i in range(100)]
with ThreadPoolExecutor(max_workers=100) as executor:
    result = list(executor.map(square, data))

print("Result (Thread):", result)

Result (Thread): [0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500, 2601, 2704, 2809, 2916, 3025, 3136, 3249, 3364, 3481, 3600, 3721, 3844, 3969, 4096, 4225, 4356, 4489, 4624, 4761, 4900, 5041, 5184, 5329, 5476, 5625, 5776, 5929, 6084, 6241, 6400, 6561, 6724, 6889, 7056, 7225, 7396, 7569, 7744, 7921, 8100, 8281, 8464, 8649, 8836, 9025, 9216, 9409, 9604, 9801]


Explanation:

In this example, we use ThreadPoolExecutor to execute a square function concurrently with up to three threads.

The square function simulates a task by squaring a number and sleeping for 1 second.

The executor.map method maps the function to data and collects the results.

This illustrates concurrent execution with threads.

## Example 2: Concurrent Execution with ProcessPoolExecutor
This code uses parallelism through ProcessPoolExecutor to execute a square function in parallel using separate processes. The square function will square each element in the data list, but with a 1-second delay to simulate a demanding task. Because max_workers=3, only three processes are running concurrently. The rest will wait until a process completes. This makes execution faster than sequential execution, but is still limited by the number of available workers. Because it uses processes (rather than threads), this approach is suitable for CPU-bound tasks that require higher performance and can efficiently utilize multiple CPU cores. The final result of the square calculation of all elements is displayed after all processes have completed.

In [11]:
from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
import time

def square(x):
    time.sleep(1)
    return x * x

data = [1, 2, 3, 4, 5, 10]
with ProcessPoolExecutor(max_workers=3) as executor:
    result = list(executor.map(square, data))

print("Result (process):", result)

BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.

In [12]:
from concurrent.futures import ProcessPoolExecutor
import time

def square(x):
    time.sleep(1)
    return x * x

if __name__ == "__main__":
    data = [1, 2, 3, 4, 5, 10]
    with ProcessPoolExecutor(max_workers=3) as executor:
        result = list(executor.map(square, data))

    print("Result (Process):", result)

BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.

Explanation:

This example is similar to the previous one, but demonstrates concurrent execution using processes instead of threads.
We use ProcessPoolExecutor to execute the square function with up to three processes.
The square function simulates the task by squaring a number and sleeping for 1 second.
The executor.map method maps the function to data, and the results are collected.
Process-based execution is suitable for tasks that require a lot of CPU resources.

The code below demonstrates the process of sequentially downloading five images from the internet using the requests module. Each image is downloaded one at a time, saved to a local file, and the total processing time is measured using the time module. This approach is sequential, so the execution time will increase as more images are downloaded.

In [13]:
import requests
import time

urls = [ 
    "https://picsum.photos/200/300", # Image 1 
    "https://picsum.photos/200/301", # Image 2 
    "https://picsum.photos/200/302", # Image 3 
    "https://picsum.photos/200/303", # Image 4 
    "https://picsum.photos/200/304", # Image 5
]

def download_image(url, index): 
    response = requests.get(url) 
    with open(f"image_{index}.jpg", "wb") as file: 
        file.write(response.content) 
    print(f"Image {index} has finished downloading.")

start_time = time.time()
for i, url in enumerate(urls): 
    download_image(url, i)

print(f"Done in {time.time() - start_time:.2f} seconds")

Image 0 has finished downloading.
Image 1 has finished downloading.
Image 2 has finished downloading.
Image 3 has finished downloading.
Image 4 has finished downloading.
Done in 9.45 seconds


This code demonstrates the process of downloading five images from the internet in parallel using ThreadPoolExecutor. By utilizing three threads, multiple images can be downloaded simultaneously, making the total execution time more efficient than a sequential approach.

In [14]:
from concurrent.futures import ThreadPoolExecutor
import requests
import time

urls = [ 
    "https://picsum.photos/200/300", 
    "https://picsum.photos/200/301", 
    "https://picsum.photos/200/302", 
    "https://picsum.photos/200/303", 
    "https://picsum.photos/200/304",
]

def download_image(url, index): 
    response = requests.get(url) 
    with open(f"image_{index}.jpg", "wb") as file: 
        file.write(response.content) 
    print(f"Image {index} has finished downloading.")

start_time = time.time()
with ThreadPoolExecutor(max_workers=3) as executor: 
    executor.map(download_image, urls, range(len(urls)))

print(f"Completed in {time.time() - start_time:.2f} seconds")

Image 2 has finished downloading.
Image 0 has finished downloading.
Image 1 has finished downloading.
Image 3 has finished downloading.
Image 4 has finished downloading.
Completed in 3.61 seconds


This code attempts to download five images in parallel using ProcessPoolExecutor, which runs tasks across multiple separate processes. While theoretically suitable for CPU-bound tasks, using requests (which cannot be picked directly) in multiprocessing can lead to errors. The primary goal is to speed up the download process by leveraging process-based parallel processing.

In [15]:
from concurrent.futures import ProcessPoolExecutor
import requests
import time

urls = [ 
    "https://picsum.photos/200/300", 
    "https://picsum.photos/200/301", 
    "https://picsum.photos/200/302", 
    "https://picsum.photos/200/303", 
    "https://picsum.photos/200/304",
]

def download_image(url, index): 
    response = requests.get(url) 
    with open(f"image_{index}.jpg", "wb") as file: 
        file.write(response.content) 
    print(f"Image {index} has finished downloading.")

start_time = time.time()
with ProcessPoolExecutor(max_workers=3) as executor: 
    executor.map(download_image, urls, range(len(urls)))

print(f"Completed in {time.time() - start_time:.2f} seconds")

Completed in 0.45 seconds


Experiments showed that downloading images one by one (sequentially) took the longest time, at around 9.45 seconds. Using ThreadPoolExecutor significantly reduced the time to around 3.61 seconds because it allows multiple downloads to run concurrently within a single process using multiple threads. Interestingly, ProcessPoolExecutor recorded the fastest time, at around 0.45 seconds, although it is generally better suited for CPU-intensive tasks (rather than I/O tasks like requests). However, these results can be inconsistent because parallel processing can be affected by cache, network, or the system environment.

Practically, for I/O-bound tasks like downloading from the internet, ThreadPoolExecutor is a safer and more efficient choice. The very fast results from ProcessPoolExecutor should be retested, as system anomalies or internal caching could be present.